### PR TITLE
Handle TLS as opaque TCP

### PR DESCRIPTION
**Pull request by ijsnellf**
HTTPS or TLS ports without associated virtual service should be treated as opaque TCP traffic. This PR fixes a bug where we were generating a TCP filter chain even when there was a corresponding TLS block, resulting in this issue: https://github.com/istio/istio/issues/6920

I've verified the behavior manually, but I need to add e2e tests still.

### CODE REVIEW

File: pilot/pkg/networking/core/v1alpha3/listener.go [hunk 1]
```diff
@@ -529,7 +529,7 @@ func (configgen *ConfigGeneratorImpl) buildSidecarOutboundListeners(env *model.E
 					// The conflict resolution is done later in this code
 				}
 
-				listenerOpts.filterChainOpts = buildOutboundTCPFilterChainOpts(node, env, configs,
+				listenerOpts.filterChainOpts = buildOutboundTCPTLSFilterChainOpts(node, env, configs,
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/listener.go]: buildSidecarOutbound..

File: pilot/pkg/networking/core/v1alpha3/tls.go [hunk 2]
```diff
@@ -108,14 +112,39 @@ func buildOutboundTCPFilterChainOpts(node *model.Proxy, env *model.Environment,
...
+	}
+
+	return out
+}
 
-		// very basic TCP (no L4 matching)
-		// break as soon as we add one network filter with no destination addresses to match
-		// This is the terminating condition in the filter chain match list
-		// TODO: rbac
+func buildOutboundTCPFilterChainOpts(node *model.Proxy, env *model.Environment, destinationIPAddress string,
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/tls.go]: sidecaroutbound..

File: pilot/pkg/networking/core/v1alpha3/tls.go [hunk 3]
```diff
@@ -108,14 +112,39 @@ func buildOutboundTCPFilterChainOpts(node *model.Proxy, env *model.Environment,
...
+	// HTTPS or TLS ports without associated virtual service will be treated as opaque TCP traffic.
+	if !hasTLSMatch {
+		clusterName := model.BuildSubsetKey(model.TrafficDirectionOutbound, "", service.Hostname, int(listenPort.Port))
+		out = append(out, &filterChainOpts{
+			destinationCIDRs: []string{destinationIPAddress},
+			networkFilters:   buildOutboundNetworkFilters(node, clusterName, destinationIPAddress, listenPort),
+		})
+	}
+
+	return out
```
Comment (rshriram) [inline pilot/pkg/networking/core/v1alpha3/tls.go]: wont this cause two tcp proxies again? coz the other function is also adding a default route

Comment (ijsnellf) [reply pilot/pkg/networking/core/v1alpha3/tls.go]: Only if there if a TCP and TLS port share the same port number, and the TLS port has no associated virtual services with a matching TLS block. That will be handled by the normal listener merging code.

In the normal case, we only add the default filter chain if the port is TCP or TLS. See:
https://github.com/istio/istio/pull/7394/files#diff-3ecbf67b650d37ef2b3af3c3bb19ba04R81
https://github.com/istio/istio/pull/7394/files#diff-3ecbf67b650d37ef2b3af3c3bb19ba04R137

**Pull request merged**